eclair-core/src/main/scala/fr/acinq/eclair/crypto/TransportHandler.scala

```diff
@@ -109,60 +108,88 @@ class TransportHandler[T: ClassTag](keyPair: KeyPair, rs: Option[BinaryData], co
-  }
-
-  when(WaitingForReadAck) {
-    case Event(ReadAck(msg: T), d: WaitingForReadAckData[T]) =>
+    case Event(ReadAck(msg: T), d: NormalData[T]) =>
       // how many occurences of this message are still unacked?
-      val remaining = d.unackedMessages.getOrElse(msg, 0) - 1
-      // if all occurences have been acked then we remove the keep from the map
-      val unacked1 = if (remaining > 0) d.unackedMessages + (msg -> remaining) else d.unackedMessages - msg
-      if (unacked1.isEmpty) {
+      val remaining = d.unackedReceived.getOrElse(msg, 0) - 1
+      // if all occurences have been acked then we remove the entry from the map
+      val unackedReceived1 = if (remaining > 0) d.unackedReceived + (msg -> remaining) else d.unackedReceived - msg
+      if (unackedReceived1.isEmpty) {
         log.debug("last incoming message was acked, resuming reading")
         connection ! Tcp.ResumeReading
-        goto(WaitingForCiphertext) using WaitingForCiphertextData(d.enc, d.dec, d.ciphertextLength, d.buffer, d.listener)
+        stay using d.copy(unackedReceived = unackedReceived1)
+      } else {
+        stay using d.copy(unackedReceived = unackedReceived1)
+      }
+
+    case Event(t: T, d: NormalData[T]) =>
+      if (d.sendBuffer.normalPriority.size + d.sendBuffer.lowPriority.size >= MAX_BUFFERED) {
+        log.warning(s"send buffer overrun, closing connection")
+        connection ! PoisonPill
+        stop(FSM.Normal)
+      } else if (d.unackedSent.isDefined) {
+        log.debug("buffering send data={}", t)
+        val sendBuffer1 = t match {
+          case _: ChannelAnnouncement => d.sendBuffer.copy(lowPriority = d.sendBuffer.lowPriority :+ t)
+          case _: NodeAnnouncement => d.sendBuffer.copy(lowPriority = d.sendBuffer.lowPriority :+ t)
+          case _: ChannelUpdate => d.sendBuffer.copy(lowPriority = d.sendBuffer.lowPriority :+ t)
+          case _ => d.sendBuffer.copy(lowPriority = d.sendBuffer.normalPriority :+ t)
+        }
+        stay using d.copy(sendBuffer = sendBuffer1)
       } else {
-        stay using d.copy(unackedMessages = unacked1)
+        val blob = codec.encode(t).require.toByteArray
+        val (enc1, ciphertext) = d.encryptor.encrypt(blob)
+        connection ! Tcp.Write(buf(ciphertext), WriteAck)
+        stay using d.copy(encryptor = enc1, unackedSent = Some(t))
       }
 
-    case Event(t: T, d: WaitingForReadAckData[T]) =>
-      val blob = codec.encode(t).require.toByteArray
-      val (enc1, ciphertext) = TransportHandler.encrypt(d.enc, blob)
-      out ! buf(ciphertext)
-      stay using d.copy(enc = enc1)
+    case Event(WriteAck, d: NormalData[T]) =>
+      def send(t: T) = {
+        val blob = codec.encode(t).require.toByteArray
+        val (enc1, ciphertext) = d.encryptor.encrypt(blob)
+        connection ! Tcp.Write(buf(ciphertext), WriteAck)
+        enc1
+      }
+      d.sendBuffer.normalPriority.dequeueOption match {
+        case Some((t, highPriority1)) =>
```

Comment:
nitpick: why not call is normalPriority1 ?

Reply:
that's a leftover, will fix